### PR TITLE
[MPLUGIN-390] Do not overwrite generate files with no content change

### PR DESCRIPTION
**Pull request by gnodet**

### CODE REVIEW

**Comment by slawekjaranowski** [comment]
Some test would be usable for regression confirm.

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/CachingOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Objects;
+
+/**
+ * Caching OutputStream to avoid overwriting a file with
+ * the same content.
+ */
+public class CachingOutputStream extends ByteArrayOutputStream
```
Comment (slawekjaranowski) [inline maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/CachingOutputStream.java:34]: I guess I've seen similar class in many place ... maybe can be put in some of share / common

**Review by slawekjaranowski — APPROVED** [review]

File: maven-plugin-plugin/src/it/mplugin-390/pom.xml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          <execution>
+            <id>touch</id>
+            <phase>generate-resources</phase>
+            <goals>
+              <goal>run</goal>
+            </goals>
+            <configuration>
+              <target>
+                <touch file="${project.build.directory}/touch.txt"/>
+                <sleep milliseconds="100"/>
```
Comment (slawekjaranowski) [inline maven-plugin-plugin/src/it/mplugin-390/pom.xml:100]: https://bugs.openjdk.java.net/browse/JDK-8177809

I don't remember where, but some time ago I have problem with lastModified .. so maybe a little longer

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/CachingOutputStream.java (hunk 1 above)
Comment (gnodet) [reply maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/CachingOutputStream.java:34]: I've introduced a similar one, but in modello (https://github.com/codehaus-plexus/modello/pull/116) and in plexus-containers (https://github.com/codehaus-plexus/plexus-containers/pull/46).  So only location would be in `plexus-utils` I suppose.  But the two classes in modello and plexus-containers are `Writer`s and this is the only `OutputStream`.

File: maven-plugin-plugin/src/it/mplugin-390/pom.xml (hunk 2 above)
Comment (gnodet) [reply maven-plugin-plugin/src/it/mplugin-390/pom.xml:100]: Raised the sleep time to 1s

File: maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/CachingOutputStream.java (hunk 1 above)
Comment (gnodet) [reply maven-plugin-tools-generators/src/main/java/org/apache/maven/tools/plugin/generator/CachingOutputStream.java]: Updated with the `CachingOutputStream` from plexus-utils.

**Review by michael-o — APPROVED** [review]
LGTM.

**Review by olamy — APPROVED** [review]

**Pull request merged**